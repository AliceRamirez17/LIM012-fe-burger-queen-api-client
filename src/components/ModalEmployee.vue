<template>
    <div class="modal">
      <div class="modal-employee">
          <label>Email</label>
          <input type="email" v-model="email">
          <label>Password</label>
          <input type="password" v-model="password">
          <label for="role">Puesto</label>
          <select v-model="role" id="role">
            <option selected>Administrador</option>
            <option >Usuario</option>
          </select>
          <button class="btn-close-modal" @click="$emit('close')"></button>
          <button class="submit-modal" @click="addFunction(userObj)">{{ button }}</button>
      </div>
    </div>
</template>

<script>
export default {
  name: 'ModalEmployee',
  data() {
    return {
      password: '',
      email: '',
      role: '',
    };
  },
  props: {
    button: String,
    editEmail: String,
    editRole: Boolean,
    addFunction: Function,
  },
  mounted () {
    if (this.button === 'Guardar cambios') {
      this.email = this.editEmail;
      if (this.editRole === true) {
        this.role = "Administrador"
      } else {
        this.role = "Usuario"
      }
    } 
  },
  computed: {
    userObj() {
      return {
        email: this.email,
        password: this.password,
        role: this.role,
      }
    }
  },
  methods: {
  }
};
</script>

<style lang="scss">
    @mixin background-img($color, $url, $size, $border, $radius) {
        background-color: $color;
        background-image: url($url);
        background-size: $size;
        border: $border;
        border-radius: $radius;
    }

    .modal {
        position: absolute;
        display: flex;
        justify-content: center;
        align-items: center;
        width: 100vw;
        height: 100vh;
        background-color: rgba($color: #000000, $alpha: 0.4);

        .modal-employee {
            position: relative;
            display: flex;
            flex-direction: column;
            justify-content: space-between;
            width: 480px;
            height: 580px;
            background-color: white;
            border-radius: 20px;
            padding: 40px;

            label {
                font-size: 24px;
                font-weight: 800;
                text-align: left;
            }

            input {
                width: 300px;
                height: 45px;
                border-radius: 10px;
                border-style: none;
                border: 1px solid black;
                outline: none;
                font-size: 22px;
                padding-left: 5px;
            }

            .btn-close-modal {
                position: absolute;
                top: 40px;
                right: 40px;
                width: 50px;
                height: 50px;
                outline: none;
                cursor: pointer;
                @include background-img(transparent,'../assets/btn-exit.svg', contain, none, 50%);
            }

            .submit-modal {
                width: 330px;
                height: 60px;
                align-self: center;
                border-radius: 10px;
                background-color: #00C7D4;
                border-style: none;
                font-size: 28px;
                font-weight: 600;
                outline: none;
                cursor: pointer;
            }

            select {
              width: 300px;
              height: 45px;
              border-radius: 10px;
              border: 1px solid black;
            }
        }
    }
</style>
