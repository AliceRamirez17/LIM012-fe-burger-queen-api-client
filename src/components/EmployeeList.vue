<template>
    <div class="employees-list">
      <div v-for="(user, index) in users" :key="index" class="each-employee">
        <h3> {{ user.email }} </h3>
        <h3 class="employee-role"> {{ user.role.admin ? 'Admin' : 'User' }} </h3>
        <button @click="showEditModal(user._id)" class="edit-employee-btn"></button>
        <button @click="showConfirmationModal(user._id)" class="delete-employee-btn"></button>
      </div>
      <confirmation-modal v-if="confirmation" @close="confirmation=false, choice={}" :addFunction="handleDeleteEmployee"/>
      <modal-employee v-if="modal" @close="modal=false"
        :editEmail="email"
        :editRole="role"
        :addFunction="handleEditEmployee"
        button="Guardar cambios"
      />
    </div>
</template>

<script>

import ConfirmationModal from './ConfirmationModal.vue';
import ModalEmployee from './ModalEmployee.vue';
import { deleteEmployee, editEmployee } from '../controllers/users';

const token = 'qwerryuipuq';

export default {
  name: 'EmployeeList',
  props: {
    users: Array,
  },
  components: {
    ModalEmployee,
    ConfirmationModal,
  },
  data() {
    return {
      modal: false,
      confirmation: false,
      choice: {},
    };
  },
  methods: {
    showEditModal(index) {
      this.modal = true;
      const userToEdit = this.users.filter(function(user) {
        return index === user._id;
      });
      this.choice = userToEdit[0];
      this.email = this.choice.email;
      this.role = this.choice.role.admin;
    },
    showConfirmationModal(index) {
      console.log('i clicked');
      this.confirmation = true;
      const modalToDelete = this.users.filter(function(user) {
        return index === user._id;
      });
      this.choice = modalToDelete[0];
    },
    handleDeleteEmployee() {
      console.log('i clicked now');
      this.confirmation = false;
      console.log(this.choice._id);
      deleteEmployee(token, this.choice._id)
        .then((resp) => {
          console.log(resp)
        })
    },
    handleEditEmployee() {
      console.log('i clicked in button edit'); 
      this.modal = false;
      editEmployee(token, this.choice, this.choice._id)
        .then((resp) => {
          console.log(resp)
        })
    },
  },
};

</script>

<style lang="scss">
  .employees-list {
      display: flex;
      width: 90%;
      overflow-y: scroll;
      grid-column: 2 / 3;
      grid-row: 2 / 3;
      flex-direction: column;
      border-radius: 20px;
      border: 1px solid black;
      box-sizing: content-box;
      margin: 40px auto;

      .each-employee {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        align-items: center;
        padding: 20px 30px;
        box-sizing: border-box;

        h3 {
          font-family: Livvic, Helvetica, Arial, sans-serif;
          font-size: 25px;
          width: 10%;
        }

        .employee-role {
          margin-left: 80px;
        }

        .edit-employee-btn, .delete-employee-btn {
          width: 50px;
          height: 50px;
          background-size: contain;
          border: none;
          background-color: transparent;
          outline: none;
          cursor: pointer;
        }

        .edit-employee-btn {
          background-color: #00C7D4;
          border-radius: 10px;
          background-image: url(../assets/btn-edit.svg);
          background-size: 70%;
          background-repeat: no-repeat;
          background-position: center;
        }

        .delete-employee-btn {
          background-image: url(../assets/btn-delete.svg);
        }
      }
    }
</style>
