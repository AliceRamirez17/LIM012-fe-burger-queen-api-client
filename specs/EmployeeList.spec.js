import { mount } from '@vue/test-utils'

import EmployeeList from '../src/components/EmployeeList.vue'
import ModalEmployee from '../src/components/ModalEmployee.vue'
import ConfirmationModal from '../src/components/ConfirmationModal.vue'

const arrUsers = [
  {
    "_id": 1,
    "email": "alice@gmail.com",
    "role": {
        "admin": true
    }
  },
  {
    "_id": 2,
    "email": "michelle@gmail.com",
    "role": {
        "admin": false
    }
  }
]

describe('Testing Employee List component', () => {
  test('prop users return a array of users', () => {
    const wrapper = mount(EmployeeList, {
      propsData: {
        users: arrUsers
      }
    })
    expect(wrapper.props().users).toBe(arrUsers)
  
    expect(wrapper.html()).toMatchSnapshot()
  })
  
  test('Modal and confirmation initializes to false', () => {
    const wrapper = mount(EmployeeList)
    expect(wrapper.vm.modal).toBe(false)
    expect(wrapper.vm.confirmation).toBe(false)
  
    expect(wrapper.html()).toMatchSnapshot()
  })
  
  test('Find components', async () => {
    const wrapper = mount(EmployeeList, {
      data() {
        return {
          modal: true,
          confirmation: true
        }
      }
    })

    expect(wrapper.findComponent(ModalEmployee).exists()).toBeTruthy()
    expect(wrapper.findComponent(ConfirmationModal).exists()).toBeTruthy()
  })
})