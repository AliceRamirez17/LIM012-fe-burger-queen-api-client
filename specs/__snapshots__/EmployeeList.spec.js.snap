// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Employee List component Modal and confirmation initializes to false 1`] = `
"<div class=\\"employees-list\\">
  <!---->
  <!---->
</div>"
`;

exports[`Testing Employee List component prop users return a array of users 1`] = `
"<div class=\\"employees-list\\">
  <div class=\\"each-employee\\">
    <h3> alice@gmail.com </h3>
    <h3 class=\\"employee-role\\"> Admin </h3> <button class=\\"edit-employee-btn\\"></button> <button class=\\"delete-employee-btn\\"></button>
  </div>
  <div class=\\"each-employee\\">
    <h3> michelle@gmail.com </h3>
    <h3 class=\\"employee-role\\"> User </h3> <button class=\\"edit-employee-btn\\"></button> <button class=\\"delete-employee-btn\\"></button>
  </div>
  <!---->
  <!---->
</div>"
`;
